src/Runner/Runner.php

```diff
@@ -89,26 +119,245 @@ public function __construct(
+                'Unable to parse server port from "%s"',
+                $server->getAddress() ?? ''
+            ));
+        }
 
-        $finder = $this->finder;
-        $finderIterator = $finder instanceof \IteratorAggregate ? $finder->getIterator() : $finder;
-        $fileFilteredFileIterator = new FileFilterIterator(
-            $finderIterator,
-            $this->eventDispatcher,
-            $this->cacheManager
+        $processPool = new ProcessPool($server);
+        $fileIterator = $this->getFileIterator();
+        $fileIterator->rewind();
+
+        $fileChunk = function () use ($fileIterator): array {
+            $files = [];
+
+            while (\count($files) < $this->parallelConfig->getFilesPerProcess()) {
+                $current = $fileIterator->current();
+
+                if (null === $current) {
+                    break;
+                }
+
+                $files[] = $current->getRealPath();
+
+                $fileIterator->next();
+            }
+
+            return $files;
+        };
+
+        // [REACT] Handle worker's handshake (init connection)
+        $server->on('connection', static function (ConnectionInterface $connection) use ($processPool, $fileChunk): void {
+            $jsonInvalidUtf8Ignore = \defined('JSON_INVALID_UTF8_IGNORE') ? JSON_INVALID_UTF8_IGNORE : 0;
+            $decoder = new Decoder($connection, true, 512, $jsonInvalidUtf8Ignore);
+            $encoder = new Encoder($connection, $jsonInvalidUtf8Ignore);
+
+            // [REACT] Bind connection when worker's process requests "hello" action (enables 2-way communication)
+            $decoder->on('data', static function (array $data) use ($processPool, $fileChunk, $decoder, $encoder): void {
+                if (ParallelAction::RUNNER_HELLO !== $data['action']) {
+                    return;
+                }
+
+                $identifier = ProcessIdentifier::fromRaw($data['identifier']);
+                $process = $processPool->getProcess($identifier);
+                $process->bindConnection($decoder, $encoder);
+                $job = $fileChunk();
+
+                if (0 === \count($job)) {
+                    $process->request(['action' => ParallelAction::WORKER_THANK_YOU]);
+                    $processPool->endProcessIfKnown($identifier);
+
+                    return;
+                }
+
+                $process->request(['action' => ParallelAction::WORKER_RUN, 'files' => $job]);
+            });
+        });
+
+        $processesToSpawn = min(
+            $this->parallelConfig->getMaxProcesses(),
+            (int) ceil($this->fileCount / $this->parallelConfig->getFilesPerProcess())
         );
+        $processFactory = new ProcessFactory($this->input);
+
+        for ($i = 0; $i < $processesToSpawn; ++$i) {
+            $identifier = ProcessIdentifier::create();
+            $process = $processFactory->create(
+                $streamSelectLoop,
+                new RunnerConfig(
```

Comment:
this is only place we create it right ?
I'm not sure of it's need, especially if we don't feed Runner with RunnerConfig, would love to understand long term usage of this (if now its limited due to BC promise?) or how we select properties that goes here (eg config vs dependency services)

Reply:
It's always better to encapsulate set of params with such class, because we can then add more params without the need to modify `ProcessFactory::create()`'s signature (and any other potential `RunnerConfig`'s consumer).